odoo/addons/base/models/res_company.py

```diff
@@ -376,24 +376,74 @@ def _get_main_company(self):
-        return (children & accessible).ids
+        while current:
+            rslt.extend((current & accessible).ids)
+            current = current.child_ids
+
+        if not rslt and self.env.su:
+            # Accessible companies will always be the same for super user when called in a cron.
+            # Because of that, the intersection between them and self might be empty. The super user anyway always has
+            # access to all companies (as it bypasses the record rules), so we return the current company in this case.
+            return self.ids
+
+        return rslt
 
     def _accessible_branches(self):
         return self.browse(self.__accessible_branches())
 
     def _all_branches_selected(self):
-        """Return whether or all the branches of the main company are selected.
+        """Return whether or all the branches of the companies in self are selected.
 
         Is ``True`` if all the branches, and only those, are selected.
-        Can be used when some actions only make sense for a whole company regardless of the
+        Can be used when some actions only make sense for whole companies regardless of the
         branches.
         """
-        return (
-            len(self.env.companies.root_id) == 1
-            and self.env.companies == self.env['res.company'].sudo().search([('id', 'child_of', self.env.company.root_id.ids)])
-        )
+        return self == self.sudo().search([('id', 'child_of', self.root_id.ids)])
+
+    def  _get_branches_with_same_vat(self, accessible_only=False):
+        """ Returns all companies among self and its sub-branches that share the same VAT number as self. An empty VAT number is considered
+        as being the same as the one of the closest parent with a VAT number.
+
+        Example:
+        - main company ; vat = 123
+            - branch 1
+                - branch 1_1
+            - branch 2 ; vat = 456
+                - branch 2_1 ; vat = 789
+                - branch 2_2
+
+        In this example, the following VAT numbers will be considered for each company:
+        - main company: 123
+        - branch 1: 123
+        - branch 1_1: 123
+        - branch 2: 456
+        - branch 2_1: 789
+        - branch 2_2: 456
+
+        :param accessible_only: whether the returned companies should exclude companies that are not in self.env.companies
+        """
+        self.ensure_one()
+
+        current = self.sudo()
+        rslt = []
+        current_vat_set = {current.vat} if current.vat else set()
+        current_strict_parents = current.parent_ids - current
+        if accessible_only:
+            candidate_branches = current._accessible_branches()
+        else:
+            candidate_branches = self.env['res.company'].sudo().search([('id', 'child_of', current.root_id.ids)])
+
+        for branch in candidate_branches:
+            if set(filter(None, (branch.parent_ids - current_strict_parents).mapped('vat'))) == current_vat_set:
```

Comment:
Isn't this more readable and doing the same?
```suggestion
            if all(c.vat == current.vat for c in branch.parent_ids - current_strict_parents):
```

Reply:
I tried, and it broke my test; it's not equivalent. 

Consider this setup
- Main company, vat = 123
--- Branch 1
------ Branch 1.1
--- Branch 2, vat = 456
------ Branch 2.1

When calling _get_branches_with_same_vat on branch 2 (with accessible_only=False), here is what happens: 
- For branch 1, (branch.parent_ids - current_strict_parents) will correspond to only itself, giving a vat set of {False}. Branch 1 will hence be accepted because {False} <= {'456', False}

- For branch 1.1, (branch.parent_ids - current_strict_parents) corresponds to branch_1 + branch_1_1, giving again a set of {False}, so branch_1_1 is also accepted

- For main company, (branch.parent_ids - current_strict_parents) gives an empty recordset, so the set will be empty. And set() <= {'456', False}, so the main company is also accepted. 


I rearrenged the code a little bit, but the filter() seems way cleaner here than checking these cases individually.